### PR TITLE
Introduce RPC class and start wiring up in Service

### PR DESCRIPTION
**Pull request by cschuet**
[RFC=0002](https://github.com/googlecartographer/rfcs/blob/master/text/0002-cloud-based-mapping-1.md)

### CODE REVIEW

File: cartographer_grpc/framework/rpc.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+namespace framework {
+
+class Service;
+class Rpc {
+ public:
+  enum class State { NEW_CONNECTION = 0, READ, WRITE, DONE };
+  struct RpcState {
+    const State state;
+    Rpc* rpc;
+    void* GetTag() { return reinterpret_cast<void*>(this); }
```
Comment (gaschler) [inline cartographer_grpc/framework/rpc.h]: Is this really needed? Can't you just pass RpcState* to a function that accepts void*?

File: cartographer_grpc/framework/service.cc [hunk 2]
```diff
@@ -16,11 +16,53 @@
...
 
+#include "glog/logging.h"
+#include "grpc++/impl/codegen/proto_utils.h"
+
 namespace cartographer_grpc {
 namespace framework {
 
-Service::Service(const std::map<std::string, RpcHandlerInfo>& rpc_handlers)
-    : rpc_handlers_(rpc_handlers) {}
+Service::Service(const std::string service_name,
```
Comment (gaschler) [inline cartographer_grpc/framework/service.cc]: const std::string*&*

File: cartographer_grpc/framework/service.cc [hunk 3]
```diff
@@ -16,11 +16,53 @@
...
+    // The 'handler' below is set to 'nullptr' indicating that we want to
+    // handle this method asynchronously.
+    this->AddMethod(new grpc::internal::RpcServiceMethod(
+        fully_qualified_method_name.c_str(), rpc_handler_info.second.rpc_type,
+        nullptr /* handler */));
+  }
+}
+
+void Service::StartServing(
+    std::vector<::grpc::ServerCompletionQueue*> completion_queues) {
```
Comment (gaschler) [inline cartographer_grpc/framework/service.cc]: const std::vector<>& ?

File: cartographer_grpc/framework/service.cc [hunk 4]
```diff
@@ -16,11 +16,53 @@
...
+    ::grpc::ServerCompletionQueue* completion_queue) {
+  switch (rpc->rpc_type()) {
+    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+      RequestAsyncClientStreaming(
+          method_index, rpc->server_context(), rpc->responder(),
+          completion_queue, completion_queue,
+          rpc->GetState(Rpc::State::NEW_CONNECTION)->GetTag());
+      break;
+    default:
+      LOG(FATAL) << "No RPC type not implemented.";
```
Comment (gaschler) [inline cartographer_grpc/framework/service.cc]: Remove "No"

File: cartographer_grpc/framework/rpc.cc [hunk 5]
```diff
@@ -0,0 +1,62 @@
+#include "cartographer_grpc/framework/rpc.h"
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/rpc.cc]: Done

File: cartographer_grpc/framework/rpc.cc [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    case State::WRITE:
+      return &write_state_;
+    case State::DONE:
+      return &done_state_;
+  }
+  LOG(FATAL) << "Never reached.";
+}
+
+ActiveRpcs::~ActiveRpcs() {
+  std::lock_guard<std::mutex> lock(mu_);
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/rpc.cc]: Done.

File: cartographer_grpc/framework/rpc.cc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+bool ActiveRpcs::Remove(Rpc* rpc) {
+  std::lock_guard<std::mutex> lock(mu_);
+  auto it = rpcs_.find(rpc);
+  if (it != rpcs_.end()) {
+    delete rpc;
+    rpcs_.erase(it);
+    return true;
+  } else {
+    return false;
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/rpc.cc]: Done.

File: cartographer_grpc/framework/rpc.h [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_GRPC_FRAMEWORK_RPC_H
+#define CARTOGRAPHER_GRPC_FRAMEWORK_RPC_H
+
+#include <memory>
+#include <mutex>
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/rpc.h]: Done.

File: cartographer_grpc/framework/rpc.h (hunk 1 above)
Comment (cschuet) [reply cartographer_grpc/framework/rpc.h]: Removed.

File: cartographer_grpc/framework/rpc.h [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  RpcState read_state_ = RpcState{State::READ, this};
+  RpcState write_state_ = RpcState{State::WRITE, this};
+  RpcState done_state_ = RpcState{State::DONE, this};
+
+  std::unique_ptr<google::protobuf::Message> request_;
+  std::unique_ptr<google::protobuf::Message> response_;
+};
+
+// This class keeps track of all in-flight RPCs for a 'Service'. Make sure that
+// all RPCs have been terminated and remove from this object before allowing to
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/rpc.h]: Done

File: cartographer_grpc/framework/rpc.h [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+// to go out of scope.
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/rpc.h]: Done.

File: cartographer_grpc/framework/rpc.h [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+class ActiveRpcs {
+ public:
+  ActiveRpcs() = default;
+  ~ActiveRpcs();
+
+  Rpc* Add(std::unique_ptr<Rpc> rpc);
+  bool Remove(Rpc* rpc);
+
+ private:
+  std::mutex mu_;
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/rpc.h]: Done.

File: cartographer_grpc/framework/service.cc (hunk 2 above)
Comment (cschuet) [reply cartographer_grpc/framework/service.cc]: Done

File: cartographer_grpc/framework/service.cc (hunk 3 above)
Comment (cschuet) [reply cartographer_grpc/framework/service.cc]: Done.

File: cartographer_grpc/framework/service.cc (hunk 4 above)
Comment (cschuet) [reply cartographer_grpc/framework/service.cc]: Done.

File: cartographer_grpc/framework/type_traits.h [hunk 12]
```diff
@@ -0,0 +1,69 @@
+/*
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/framework/type_traits.h]: Ack, but this way you can see it in action :)

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**